source/agora/common/BlockStorage.d

```diff
@@ -0,0 +1,277 @@
+
+        this.block_positions[block.header.height] = pos;
+
+        // write to memory
+        foreach (idx, ref e; serialized_block)
+            this.file[pos + idx] = e;
+
+        return true;
+    }
+
+    /***************************************************************************
+
+        Read block from the file.
+
+        Params:
+            block = `Block` to read
+
+        Returns:
+            Returns true if success, otherwise returns false.
+
+    ***************************************************************************/
+
+    public bool readBlock (ref Block block, size_t height)
+    {
+        if (height >= this.block_positions.length)
+            return false;
+
+        const size_t next_height = height + 1;
+        this.map();
+
+        const size_t from = this.block_positions[height];
+        size_t to;
+
+        if (next_height >= this.block_positions.length)
+            to = file.length;
+        else
+            to = this.block_positions[next_height];
+
+        assert(from < file.length);
+        assert(from < to);
+
+        block = deserializeBlock(cast(ubyte[])this.file[from .. to]);
+
+        return true;
+    }
+}
+
+/// Create and store 1000 blocks and read again. And read 100 random.
+unittest
+{
+    import agora.common.crypto.Key;
+    import agora.common.Data;
+    import agora.common.Hash;
+    import agora.common.Transaction;
+    import agora.consensus.Genesis;
+
+    import std.algorithm.comparison;
+
+
+    string path = buildPath(getcwd, ".cache");
+    if (!path.exists)
+        mkdir(path);
+
+    BlockStorage storage = new BlockStorage(path);
+
+    KeyPair[] key_pairs = [
+        KeyPair.random, KeyPair.random, KeyPair.random, KeyPair.random,
+        KeyPair.random, KeyPair.random, KeyPair.random, KeyPair.random
+    ];
+    Block[] blocks;
+
+    blocks ~= getGenesisBlock();
+    storage.saveBlock(blocks[$ - 1]);
+
+    auto gen_key_pair = getGenesisKeyPair();
+    Transaction last_tx = blocks[$ - 1].txs[$-1];
+    Hash gen_tx_hash = hashFull(last_tx);
+
+    Transaction tx;
+
+    // save
+    foreach (idx; 1 .. 1_000)
+    {
+        tx = Transaction(
+            [
+                Input(gen_tx_hash, 0)
+            ],
+            [
+                Output(Amount(1_000), key_pairs[idx % 8].address)
+            ]
+        );
+        tx.inputs[0].signature = gen_key_pair.secret.sign(hashFull(tx)[]);
+        blocks ~= makeNewBlock(blocks[$ - 1], [tx]);
+        storage.saveBlock(blocks[$ - 1]);
+    }
+
+    // load
+    Block[] loaded_blocks;
+    loaded_blocks.length = 1_000;
+    foreach (idx; 0 .. 1_000)
+        storage.readBlock(loaded_blocks[idx], idx);
+
+    // compare
+    assert(equal(blocks, loaded_blocks));
+
+    // checks data file
+    assert(storage.getFileName().exists);
+
+    // test of random access
+    import std.random;
+
+    auto rnd = rndGen;
+
+    Block random_block;
+    foreach (idx; 0 .. 100)
+    {
+        auto height = uniform(0, 999, rnd);
```

Comment:
If you want to test with all the heights, but in random order, then you can do this:

```
foreach (height; iota(999).randomCover(rnd))
{
}
```

Reply:
It's a very good function that you haven't seen anywhere else, and it's amazing. Thank you.